### PR TITLE
[promptflow] feat: better format output for `pf flow validate`

### PR DESCRIPTION
**Pull request by elliotzh**
# Description

Vendor `_validation` from `azure-ai-ml` to better format output of `pf flow validate`.

+ return a ValidationResult instead of a simple dict in operation `validate`
+ always output a json dict. return this for validation succeed:
```json
{
    "result": "Succeeded"
}
```
+ better format output for failure:
```json
{
  "result": "Failed",
  "errors": [
    {
      "message": "Missing data for required field.",
      "path": "inputs.url.type",
      "value": null,
      "location": "C:\\PyCharmProjects\\promptflow-core\\src\\promptflow\\tests\\test_configs\\flows\\web_classification_invalid\\flow.dag.yaml#line 2"
    },
    {
      "message": "Missing data for required field.",
      "path": "outputs.category.type",
      "value": null,
      "location": "C:\\PyCharmProjects\\promptflow-core\\src\\promptflow\\tests\\test_configs\\flows\\web_classification_invalid\\flow.dag.yaml#line 5"
    },
    {
      "message": "Meta file not found, path 'summarize_text_content.jinja2', type 'llm'.",
      "path": "nodes.1.source.path",
      "value": "summarize_text_content.jinja2",
      "location": "C:\\PyCharmProjects\\promptflow-core\\src\\promptflow\\tests\\test_configs\\flows\\web_classification_invalid\\flow.dag.yaml#line 22"
    },
    {
      "message": "Meta file not found, path 'summarize_text_content.jinja2', type 'llm'.",
      "path": "node_variants.summarize_text_content.variants.variant_0.node.source.path",
      "value": "summarize_text_content.jinja2",
      "location": "C:\\PyCharmProjects\\promptflow-core\\src\\promptflow\\tests\\test_configs\\flows\\web_classification_invalid\\flow.dag.yaml#line 96"
    }
  ]
}
```

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/_vendor/_validation/core.py [hunk 1]
```diff
@@ -0,0 +1,524 @@
+# ---------------------------------------------------------
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_vendor/_validation/core.py:1]: no need to put this as vendor? We might customize this for promptflow freely

File: src/promptflow/promptflow/_sdk/_vendor/_validation/core.py [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+        """
+        return cls(
+            yaml_path=yaml_path,
+            message=message,
+            error_code=error_code,
+        )
+
+
+class ValidationResult(object):
+    """Represents the result of job/asset validation.
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_vendor/_validation/core.py:87]: remove job/asset ?

File: src/promptflow/promptflow/_sdk/_vendor/_validation/core.py [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+        """
+        Return all messages of errors in the validation result.
+
+        :return: A dictionary of error messages. The key is the yaml path of the error, and the value is the error
+            message.
+        :rtype: dict
+
+        .. admonition:: Example:
+
+            .. literalinclude:: ../../../../samples/ml_samples_misc.py
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_vendor/_validation/core.py:109]: update this

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_local_operations.py [hunk 4]
```diff
@@ -281,15 +282,21 @@ def test_flow_validation_failed(self, pf) -> None:
...
+            "nodes.1.source.path",
+        ]:
+            assert "Meta file not found" in error_messages.pop(yaml_path, "")
+
+        assert error_messages == {
+            "inputs.url.type": "Missing data for required field.",
+            "outputs.category.type": "Missing data for required field.",
         }
 
+        assert "line 22" in repr(validation_result)
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_flow_local_operations.py:298]: what's the str repr for a success validate result?

File: src/promptflow/promptflow/_sdk/operations/_flow_operations.py [hunk 5]
```diff
@@ -444,73 +431,58 @@ def _resolve_additional_includes(cls, flow_dag_path: Path) -> Iterable[Path]:
         else:
             yield flow_dag_path
 
-    def validate(self, flow: Union[str, PathLike], *, raise_error: bool = False, **kwargs) -> dict:
+    def validate(self, flow: Union[str, PathLike], *, raise_error: bool = False, **kwargs) -> ValidationResult:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_flow_operations.py:434]: expose the ValidationResult class

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_local_operations.py (hunk 4 above)
Comment (elliotzh) [reply src/promptflow/tests/sdk_cli_test/e2etests/test_flow_local_operations.py:298]:
```json
{
    "result": "Succeeded"
}
```

File: src/promptflow/promptflow/_sdk/operations/_flow_operations.py [hunk 6]
```diff
@@ -444,73 +431,58 @@ def _resolve_additional_includes(cls, flow_dag_path: Path) -> Iterable[Path]:
...
         """
         Validate flow.
 
         :param flow: path to the flow directory or flow dag to export
         :type flow: Union[str, PathLike]
         :param raise_error: whether raise error when validation failed
         :type raise_error: bool
-        :return: dict of validation result
-        :rtype: dict
+        :return: a validation result object
```
[the remark this answers is not in the document]
Comment (elliotzh) [reply src/promptflow/promptflow/_sdk/operations/_flow_operations.py:442]: how can I verify this?

[the remark this answers is not in the document]
Comment (wangchao1230) [reply src/promptflow/promptflow/_sdk/operations/_flow_operations.py:442]: + @0mza987 here for help

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 7]
```diff
@@ -105,3 +115,66 @@ def _get_local_connections(cls, executable):
                 # ignore when connection not found since it can be configured with env var.
                 raise Exception(f"Connection {n!r} required for flow {executable.name!r} is not found.")
         return result
+
+
+class ProtectedFlow(Flow, SchemaValidatableMixin):
```
Comment (D-W-) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:120]: Add some docsting for this class?

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 8]
```diff
@@ -105,3 +115,66 @@ def _get_local_connections(cls, executable):
...
+    def __init__(
+        self,
+        code: str,
+        **kwargs,
+    ):
+        super().__init__(code=code, **kwargs)
+
+        self._flow_dir, self._dag_file_name = self._get_flow_definition(self.code)
+
+    @property
```
Comment (D-W-) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:130]: Seems some properties can be put in Flow class?

Comment (elliotzh) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:130]: Hide all of them for now to avoid impacting public interface

Comment (elliotzh) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:124]: Explained in docstring

File: src/promptflow/promptflow/_sdk/entities/_flow.py (hunk 7 above)
Comment (elliotzh) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:105]: updated.

**Review by D-W- — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**